File: src/main/java/com/machfour/macros/cli/modes/Export.java
package com.machfour.macros.cli.modes;

import com.machfour.macros.cli.CommandImpl;
import com.machfour.macros.core.MacrosEntity;
import com.machfour.macros.core.Table;
import com.machfour.macros.objects.*;
import com.machfour.macros.storage.CsvExport;
import com.machfour.macros.storage.MacrosDataSource;
import com.machfour.macros.util.FileUtils;

import java.io.FileWriter;
import java.io.IOException;
import java.io.Writer;
import java.sql.SQLException;
import java.util.List;



public class Export extends CommandImpl {
    private static final String NAME = "export";
    private static final String USAGE = String.format("%s %s [output dir]", config.getProgramName(), NAME);

    public Export() {
        super(NAME, USAGE);
    }

    @Override
    public void printHelp() {
        out.println("Exports complete CSV data (foods and servings) from the database.");
        out.println("Please specify the containing directory. It will be created if it doesn't exist.");
    }

    private <M extends MacrosEntity<M>> void exportTable(MacrosDataSource db, String outDir, Table<M> t)
            throws SQLException, IOException {
        out.println("Exporting " + t.getName() + " table...");
        String outCsvPath = FileUtils.joinPath(outDir, t.getName() + ".csv");
        try (Writer outCsv = new FileWriter(outCsvPath)) {
            CsvExport.exportTable(t, outCsv, db);
        }
    }

    @Override
    public int doAction(List<String> args) {
        if (args.contains("--help")) {
            printHelp();
            return 0;
        }

        String outputDir = args.size() >= 2 ? args.get(1) : config.getDefaultCsvOutputDir();
        MacrosDataSource ds =  config.getDataSourceInstance();

        try {
            exportTable(ds, outputDir, Food.table());
            exportTable(ds, outputDir, NutritionData.table());
            exportTable(ds, outputDir, Serving.table());
            exportTable(ds, outputDir, Ingredient.table());
            exportTable(ds, outputDir, Meal.table());
            exportTable(ds, outputDir, FoodPortion.table());
        } catch (SQLException | IOException e) {
            out.println();
            err.printf("Exception occurred (%s). Message: %s\n", e.getClass(), e.getMessage());
            return 1;
        }

        out.println();
        out.println("Export completed successfully");
        return 0;
    }
}

File: src/main/kotlin/com/machfour/macros/cli/modes/Export.kt
package com.machfour.macros.cli.modes
import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.printlnErr
import com.machfour.macros.core.MacrosConfig
import com.machfour.macros.csv.exportFoodData
import com.machfour.macros.csv.exportServings
import com.machfour.macros.schema.FoodTable
import com.machfour.macros.schema.ServingTable
import com.machfour.macros.sql.Table
import com.machfour.macros.util.currentTimeString
import java.io.FileWriter
import java.io.IOException
import java.sql.SQLException

class Export(config: MacrosConfig): CommandImpl(NAME, USAGE, config) {
    companion object {
        private const val NAME = "export"
        private const val USAGE = "Usage: $programName $NAME [output.zip]"
    }

    override fun printHelp() {
        println(
            "Exports user-added foods and servings to CSV files which can be re-imported when setting up" +
            "a new database. However, some information such as meal records and creation/modification times" +
            "is not preserved."
        )
    }

    private fun getExportCsvName(table: Table<*>) = "${table.name}-${currentTimeString()}.csv"

    override fun doAction(args: List<String>): Int {
        if (args.contains("--help")) {
            printHelp()
            return 0
        }

        val foodKeyCol = FoodTable.INDEX_NAME
        val outputFoodCsvName = getExportCsvName(FoodTable)
        val outputServingCsvName = getExportCsvName(ServingTable)
        try {
            println("Exporting foods to $outputFoodCsvName")
            FileWriter(outputFoodCsvName).use {
                exportFoodData(config.database, it, foodKeyCol)
            }

            println("Exporting servings to $outputServingCsvName")

            FileWriter(outputServingCsvName).use {
                exportServings(config.database, it, foodKeyCol, emptySet())
            }

        } catch (e: SQLException) {
            return handleException(e)
        } catch (e: IOException) {
            return handleException(e)
        }

        println()
        println("Export completed successfully")
        return 0
    }

    private fun handleException(e: Exception): Int {
        println()
        printlnErr("Exception occurred (${e.javaClass}). Message: ${e.message}")
        return 1
    }

}